Page Component Page - Alternate to #2471 (#2508)
grommet/hpe-design-system: aries-site/src/pages/components/page.mdx

import { Example } from '../../layouts';
import { 
  PageBackGroundExample,
  PageExample, 
  PageFullExample,
  PageNarrowExample,
  SinglePageContent,
  PageWideExample 
} from '../../examples';

Page makes it easy to implement [page layouts](/templates/page-layouts) uniformly across HPE 
applications. Two components, Page and PageContent, work together to ensure HPE theme-defined 
padding, horizontal alignment, max-width, and responsive behaviors are applied in a consistent, 
turn-key manner.

<Example 
  docs="https://v2.grommet.io/page?theme=hpe"

>   code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/content-layouts/ResponsiveContentLayoutExample.js"

Reviewer: Since this example is more visual should we leave out the code? I know in other places we have just the `Preview` open to opinions just a thought

Author: Another good question. I was thinking it would be valuable to be able to inspect how that was accomplished, but we provide the ability to do that elsewhere, so not necessarily needed here. I'm leaning towards your instinct to remove. @taysea thoughts?

Reviewer 2: Agree, think we can remove for this example.

Also side note: noticing some funny behavior when I scroll:
<img width="1515" alt="Screen Shot 2022-04-06 at 10 54 53 AM" src="https://user-images.githubusercontent.com/12522275/162037985-0f5bb1bd-c2eb-4fb8-86b4-90a08e182ee8.png">

Author: I've fixed that.

Author: done

  showResponsiveControls={['fullScreen']}
  // height, width, and screenContainer.scale value chosen to best visually 
  // represent the example
  screenContainer={{scale: 0.85}}
  width="100%"
>
  <PageExample />
</Example>

## Page and PageContent Working Together
Page is the outer layout component and provides the desired layout context. 
Page’s `kind` determines which theme-defined alignment, padding, and min/max-widths 
are applied to its PageContent children.

As a child of Page, PageContent consumes and adheres to the 
layout styling provided by its parent Page.

```
<Page kind="wide">
  <PageContent>

    /* All contents within me will respect the 
     * max-width, alignment, and padding specified 
     * by Page's context. */
     
  </PageContent>
  

>   /* Contents outside of PageContent are not bound by 

Reviewer: should we add that everything should be in the `PageContent` by reading this teams may think hey if I want something to expand I'll just move it out of `PageContent` however that's not our recommendation correct?

Author: 🤔 ... not sure.

Trying to explain the mechanics / provide the reader a mental model for how Page and PageContent relate. 

I'm not sure if that sort of caveat is valuable or not. @taysea thoughts?

Reviewer 2: Hmmm.. I could go both ways on this. Since we are generally encouraging people to put page content inside of the PageContent, I'm slightly inclined to leave the "Contents outside ..." out for now.

However, designs using [differing background colors](#full-width-background-color) to delineate page 
sections may accomplish this by applying a PageContent per desired section.

## Example use cases

### Simple
All content is placed within a single PageContent component.

<Example
  code={[
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/page/SinglePageContent.js",
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/FilterControls/FilterControls.js",
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/FilterControls/FiltersContext.js",
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/PageHeader/PageHeader.js",
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/data/mockData/trainingDataSets.json",
  ]}
  // height, width, and screenContainer.scale value chosen to best visually 
  // represent the example
  width="100%" 
  showResponsiveControls={['fullScreen']}
  screenContainer={{ scale: 0.6 }}
>

>   <SinglePageContent />

Reviewer 2: I'm finding the amount of dashed boxes confusing. It feels like the dashed boxes should be used to highlight the components specific to this page (Page and PageContent), so I would only expect to have 2 dashed boxes.

The dashed boxes relating to how content is built up seems too tangential to the topic.

<img width="1536" alt="Screen Shot 2022-04-05 at 2 47 04 PM" src="https://user-images.githubusercontent.com/12522275/161855689-e7646dc8-36e3-490e-991a-d309499f02ed.png">

Author: I agree. Presented option for reaction. Will only apply to Page and PageContent.

Reviewer: I also agree I liked the dashed line on the first example made it easier to follow

Author: done

</Example>

### Full-width background color
Use cases where an application needs a background to span the entire width of a page section 
is achieved by setting `background={{ fill: "horizontal", ...additional properties }}` on 
PageContent.

This treatment is often seen on marketing sites where backgrounds help visually 
block content sections. This might also be seen in a context such as an app-store catalog where 
the hero section might have an image spanning.

<Example
  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/page/PageContentExample.js"
  // height, width, and screenContainer.scale value chosen to best visually 
  // represent the example
  width="100%" 
  showResponsiveControls={['fullScreen']}
  screenContainer={{ scale: 0.57 }}
>

>   <PageBackGroundExample />

Reviewer 2: The background color with the gradient looks a bit abrupt to me. I'd probably prefer just the ink splash on the top left and then white for the rest of the background.

If this is 2 PageContents, should we delineate both with a dashed line?

<img width="1399" alt="Screen Shot 2022-04-05 at 2 47 47 PM" src="https://user-images.githubusercontent.com/12522275/161855760-37877116-c54b-4261-9f64-3e17b8852a24.png">

Reviewer 2: Also, we should change the content in one of the sections so it doesn't repeat.

Author: See the "ready for review" section of the PR description. This example is not ready for review.

Reviewer 2: Thanks, ignore this comment then.

Author: this is ready for review now

</Example>

> ## Variants 

Reviewer: any reason the Simple and Background Example are not here under Variants? 
Or are we only explicitly having the three `kinds` under variants?

Author: mmmm... good question. In my mind I see a distinction between what I is trying to be communicated in the two sections, but I am having trouble articulating... which is a signal that I need to develop greater clarity in my thought, or that there isn't a justifiable distinction.

Let me attempt:
- **Variants** - I view variants as various flavors of the same thing (e.g. chocolate, vanilla, strawberry are variants of ice creams, or hamburger, fried chicken, chef salad are variants of menu items). So when I look at the "variants" section, I see the purpose of the section as presenting a menu of options I can order from.
- **Example use cases** - My intent for this section was more to try to align to or showcase the possibilities of experiences someone might want to create. Following the food analogy, what type of meal experience would you like to create? A quick lunch? A celebratory meal with friends? A romantic date night? These are the things I am trying to accomplish, experience, feel, etc. We can use our menu of options (variants) to create this experience (implement this use case). We can grab white table cloth, fine dishes, candle-light, farm-to-table ingredients to create a romantic date night; we can grab kind="wide" + background props to create a rich marketing site.

I feel like presenting types of experiences and also presenting a menu of options are both valuable. Now my questions are:
- Was this accomplished? Was it done well?
- How comprehensive should our variants be? Should `background` be part of the variants? I'm leaning towards background is a Box prop and Box should do that work, whereas `kind` is specific to page.

End of my rambling thoughts for now...

Reviewer 2: It's a good question -- I think initially we tried to have "Variants" be prop related, then in some cases (templates especially) it's more pattern focused. Maybe something for us to revisit in future. I am okay with the current differentiation of variants/use cases

### Wide
"Wide" is Page's default `kind`. The vast majority of pages in HPE applications will 
likely use Page's "wide" variant. This page layout ensures consistent content alignment 

> and padding from section to section, page to page, and app to app.

Reviewer 2: Wondering about the last section because all page kinds should ensure consistent padding in these scenarios.

Author: section = sentence?

Reviewer 2: Yes, sentence.

Author: removing

See [page layouts](/templates/page-layouts#page-kinds) for specifications 
and use-case guidance.

<Example 
  showResponsiveControls={['fullScreen']}
  docs="https://v2.grommet.io/page?theme=hpe"
  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/content-layouts/ResponsiveContentLayoutExample.js"
  // height, width, and screenContainer.scale value chosen to best visually 
  // represent the example
  height="650px"
  width="100%"
  screenContainer={{ scale: 0.57 }}
>
  <PageWideExample  />
</Example>

### Narrow

> For use cases where the desire is to provide a user with singular, focussed content or objective, 

Reviewer 2: ```suggestion
Narrow is for use cases where the desire is to provide a user with singular, focused content or objective, 
```

> such as editing field values of an object, or stepping through a [wizard-based task flow](/templates/wizard).

Reviewer 2: ```suggestion
such as editing field values of an object or stepping through a [wizard-based task flow](/templates/wizard).
```

> A "narrow" page layout limits the page content's width and positions the content at the center 

Reviewer 2: ```suggestion
A "narrow" page layout limits the page content's width to `large` and positions the content at the center 
```

Reviewer 2: Trying to differentiate from "Wide" since wide also limits the width of the page.

of the page.

See [page layouts](/templates/page-layouts#page-kinds) for specifications 
and use-case guidance.

<Example 
  showResponsiveControls={['fullScreen']}
  docs="https://v2.grommet.io/page?theme=hpe"
  code="https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/content-layouts/ResponsiveContentLayoutExample.js"
  // height, width, and screenContainer.scale value chosen to best visually 
  // represent the example
  height="650px"
  width="100%"
  screenContainer={{ scale: 0.57 }}
>
  <PageNarrowExample  />
</Example>
  

### Full
"Full" is helpful in situations like presenting a monitoring dashboard filled with a 

> collection of key performance indicators, or an interface allowing a user to browse 

Reviewer 2: ```suggestion
collection of key performance indicators or an interface allowing a user to browse 
```

<Example 
  docs="https://v2.grommet.io/page?theme=hpe"
  showResponsiveControls={['fullScreen']}
  // height, width, and screenContainer.scale value chosen to best visually 
  // represent the example
  screenContainer
  width="100%"
>
  <PageExample />
</Example>

## Page and PageContent Working Together
Page is the outer layout component and provides the desired layout context. 
Page’s `kind` determines which theme-defined alignment, padding, and min/max-widths 
are applied to its PageContent children.

As a child of Page, PageContent consumes and adheres to the 
layout styling provided by its parent Page.

```
<Page kind="wide">
  <PageContent>

    /* All contents within me will respect the 
     * max-width, alignment, and padding specified 
     * by Page's context. */
     
  </PageContent>
  
  /* Contents outside of PageContent are not bound by 
   * Page's context. */

</Page>
```

<br />

In most cases, all content on a page can be wrapped in a [single PageContent component](#simple). 
However, designs using [differing background colors](#full-width-background-color) to delineate page 
sections may accomplish this by applying a PageContent per desired section.

> ## Example use cases

Reviewer 3: ~Example use cases~ Remove this section title

Author: done

> ### Simple

Reviewer 3: To make the story flow smoother, how about we keep it simple and say:
"Example: Content within a single PageContent"

"Simple" throws me off a little bit because I start viewing it as a variant

Reviewer: I would agree here the `simple` makes me think this should be under the variants

Author: done

All content is placed within a single PageContent component.

<Example
  code={[
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/page/SinglePageContent.js",
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/FilterControls/FilterControls.js",
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/FilterControls/FiltersContext.js",
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/templates/PageHeader/PageHeader.js",
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/data/mockData/trainingDataSets.json",
  ]}
   github="https://github.com/grommet/hpe-design-system/blob/master/aries-site/src/examples/components/page/SinglePageContent.js"
  // height, width, and screenContainer.scale value chosen to best visually 
  // represent the example
  width="100%" 
  showResponsiveControls={['fullScreen']}
  screenContainer={{ scale: 0.6 }}
>
  <SinglePageContent />
</Example>

> ### Full-width background color

Reviewer 3: Same recommendation as above:

What about just saying:
"Example: Page with full-width background"

Author: done

Use cases where an application needs a background to span the entire width of a page section 
is achieved by setting `background={{ fill: "horizontal", ...additional properties }}` on 
PageContent.

This treatment is often seen on marketing sites where backgrounds help visually 
block content sections. This might also be seen in a context such as an app-store catalog where 
the hero section might have an image spanning.

<Example
  code={[
    "https://raw.githubusercontent.com/grommet/hpe-design-system/master/aries-site/src/examples/components/page/PageBackGroundExample.js",
  ]}
  github="https://github.com/grommet/hpe-design-system/blob/master/aries-site/src/examples/components/page/PageBackGroundExample.js"
  // height, width, and screenContainer.scale value chosen to best visually 
  // represent the example
  width="100%" 
  showResponsiveControls={['fullScreen']}
  screenContainer={{ scale: 0.57 }}
>
  <PageBackGroundExample />
</Example>

## Variants 

### Wide
"Wide" is Page's default `kind`. The vast majority of pages in HPE applications will 
likely use Page's "wide" variant.

> See [page layouts](/templates/page-layouts#page-kinds) for specifications 

Reviewer 3: Should we reference content layouts too?

So something like:
"See [Page Layouts](https://design-system.hpe.design/templates/page-layouts) for specifications and [Content Layouts](https://design-system.hpe.design/templates/content-layouts?q=content) for templates and guidance on responsive design."

Author: done
